### DISCUSSION TITLE
Opening PDF on a network share (samba) is slow

### DISCUSSION THREAD

**Discussion by laspinacristian** [sumatrapdfreader/sumatrapdf#3963] [Q&A]
Hello 😃 
I'm experiencing slow startup times for sumatra when opening files stored on a samba share.
If I move the file to my PC then it opens really fast. The size of PDFs i'm testing is very small, about 80KB or less.

Any idea on how I can speed things up without moving files from the network share to my local drive?
Thank you

**Comment by GitHubRulesOK** [comment]
Network comms is inherently slow for file handling. When I started working on remote files it could take an hour for a document to arrive. Due to PDF two way use of  PDF indexed rendering (constantly reading tail of file) it can be slower than more conventional gulp whole file from disk into memory.
I am surprised it is noticeable on very very small files, I would have expected a local cache would be avoiding to and fro.

**Reply by laspinacristian** [reply]
Thank you for your feedback.

Yes it is noticeable slower. It opens instantly when the file is in my local drive (<0.5s) while it take more than 2s when file is on shared folder. Strange thing is that with adobe reader the opening speed between local and network files is almost the same, so I thought this was a Sumatra problem.

Not a tragedy when you open just a file, but I was closing and opening a lot of files in a shared network this delay became annoying very soon :)

**Reply by GitHubRulesOK** [reply]
The delay is different for different systems but the steps are fairly similar

Post a message with filename request, (message to nearest server on Mars takes 22.5 Minutes at the speed of light)
Server seeks file name and verifies it exists and may send back just that message  50 Minutes just to be told there are no such files on Mars)

Another cycle simply to confirm "OK just send it in chunks regardless of  quality"
if there are quality fails I will mail you another request to replace that packet.

Several ACK nowledgements later 
OK stop sending I got it here

So there is a form of latency and the transport protocol needs to be optimal between the 2 parties. (sending shuttles of books to Mars is way slower than eBooks at light speed, but feels better in the hand )

**Reply by GitHubRulesOK** [reply]
IT is possible SumatraPDF is calling server more often than Adobe so it is worth starting a clean session then open one file observing delay then using CTRL+K with Show Log see if the log shows unusual contents.

**Reply by laspinacristian** [reply]
```
ReadRegStrTemp(HKEY_LOCAL_MACHINE, Software\Microsoft\Windows\CurrentVersion\Uninstall\SumatraPDF, InstallLocation) => 'C:\Program Files\SumatraPDF'
InstallCrashHandler crashDumpPath: 'C:\Users\cristian.laspina_ker\AppData\Local\SumatraPDF\crashinfo\sumatrapdfcrash.dmp'
crashFilePath: 'C:\Users\cristian.laspina_ker\AppData\Local\SumatraPDF\crashinfo\sumatrapdfcrash.txt'
symDir: 'C:\Users\cristian.laspina_ker\AppData\Local\SumatraPDF\crashinfo'
Starting SumatraPDF 3.5.2, GetCommandLineW():
"C:\Program Files\SumatraPDF\SumatraPDF.exe" "\\192.168.30.208\share\scansioni\20230217115056.pdf"
OpenEmbeddedFilesArchive: opened archive
LoadSettings() took 0.36 ms
ParseTranslationsTxt: 17632 lines, nStrings: 372
CreateControllerForEngineOrFile: '\\192.168.30.208\share\scansioni\20230217115056.pdf', 1 pages
LoadDocument: 535.46 ms, 1 pages for '\\192.168.30.208\share\scansioni\20230217115056.pdf'
DisplayModel::BuildPagesInfo started
DisplayModel::BuildPagesInfo took 0.01 ms
CanViewWithKnownExternalViewer cmd: 294, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 295, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 296, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 297, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 298, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 299, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 300, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 301, !ev || ev->exeFullPath == nullptr
CanViewWithKnownExternalViewer cmd: 302, !pos
CanViewWithKnownExternalViewer cmd: 303, !ev || ev->exeFullPath == nullptr
Slow rendering: 138.53 ms, page: 1 in '\\192.168.30.208\share\scansioni\20230217115056.pdf'
StartMonitoringDirForChangesAPC() C:\Users\cristian.laspina_ker\AppData\Local\SumatraPDF
```

Strangely enough it says that load document is only 500ms. But I can tell for sure that the moment I doubleclick file to the moment the Sumatra window appears it took about 2s

**Reply by GitHubRulesOK** [reply]
The time SumatraPDF shows is from the bang  of the starters gun it is able to load the file in under one second same as Local. or same as one that took 75 minutes getting backwards and forwards again to Mars. the external countdown time to launch on Earth is unknown.

A process monitor (procmon) or packet monitor on the network is probably the only way to decide why there is such a significant difference but this is the same problem as the 8 seconds for one user and 1 second for another. see https://github.com/sumatrapdfreader/sumatrapdf/issues/3947

**Reply by laspinacristian** [reply]
Thanks a lot for all the explanation.

Anyway I'm supposing that this latency isn't 'overhead' caused by the OS, other installed software and/or network.
As suggested by the op of #3947 I tried v3.1.2 and it works really fast. no delay at all.

FYI, I tried 3.5.2 in 2 different environment with 2 different PCs both standalone and installer version.
After reading #3947 I tried portable versions of both 3.5.2 and 3.1.2 on the same machine and the difference is big.

3.1.2 is pretty dated. Don't know in which version the issue appear

**Reply by GitHubRulesOK** [reply]
3.1.2 is long dated because it was perhaps the most stable of all versions (for a very long time) and there was a major code shift over more to MuPDF in 3.2 
if you can prove what is repeatable between slow and fast loading it in your environment it may be possible to debug why but the point is that SumatraPDF code should be allowed to run with external data equally soon without delay if there is no other interference or overload. so I have installed and portable on same machine
Acrobat to open a 600 KB file 8 seconds
SumatraPDF installed same file under 1 second
Drag and drop same file on portable a bit slower as it launches history so total say 3 seconds, 
clear history its under one second after drop on exe.

**Reply by laspinacristian** [reply]
It is not my environment that is slowing this operation.
It's something related to Sumatra code or one of its dependecies.

This occour only when opening files located on a network share, not when opening files stored in my PC hard drive.

If we want to make comparison i can say that Sumatra slow 3.5.2 is so slow to open **file on a network** that even Adobe is faster.
But the scope is not to compare Sumatra speed to other PDF reader, the scope is to understand what is the cause and if it's possible to revert to the old performance. 
![Animazione](https://github.com/sumatrapdfreader/sumatrapdf/assets/31296728/fbd3bb47-b2e6-4785-b3dd-338cffcc6321)

**Reply by GitHubRulesOK** [reply]
@laspinacristian  Very good example,
Personally I would have preferred to see 3.12 opened first instantly, closely followed by same slower file from recent as that would indicate it is not cached between the 2 openings. So try it the other way round.
 
@kjk there seems to be some significant difference  in the methods used to request a file as shown above, do you want to elevate to an issue or simply say newer VS code? is different in behaviours ? above suggests there is a core difference.

**Reply by laspinacristian** [reply]
Here is a new example loading file on 3.1.2 first.
Both versions are installed as portable. in the settings files for both executable I made this changes:

```
RememberOpenedFiles = false
RememberStatePerDocument = false
RestoreSession = false 
CheckForUpdates = false
FileStates [
]
SessionData [
]
```
`sumatrapdfcache folder was also deleted`

![Animazione2](https://github.com/sumatrapdfreader/sumatrapdf/assets/31296728/c0f075ff-bdbe-4782-82a0-2a00cffc65d6)

**Reply by GitHubRulesOK** [reply]
Thank that shows a difference I can not currently find a samba drive only another remote windows drive so my tests are not comparable however, for contrast I am using portable 32 bit versions  of 3.1.2 and 3.5.2 so my case deliberatly using bigger files of several MB they will be slow however loding i 3.1.2 and then loading in 3.5.2 neither seems noticeably faster, both now take about 2 seconds

Same with 250 MB (over the 32 MB threshold lock) both files naturally take longer to transport but  both appear around 3 seconds after call by drag and drop, switching to call by name from standing start

```
3.1.2>time <nul & echo SumatraPDF.exe -exit-when-done -stress-test "V:\Public\250MBtest - FIXED.pdf" 1x > temp.cmd &&echo "cmd /r time <nul" >>temp.cmd &&call temp

The current time is: 20:46:58.52
return result            20:47:25.62
difference                         27 seconds

3.52>time <nul & echo SumatraPDF-3.5.2-32.exe -exit-when-done -stress-test "V:\Public\250MBtest - FIXED.pdf" 1x > temp.cmd &&echo "cmd /r time <nul" >>temp.cmd &&call temp

The current time is: 20:44:50.76
return result            20:45:07.21
difference                         17 seconds

check 3.5.2 again
The current time is: 20:52:36.73
The current time is: 20:52:42.28
difference                          7 seconds 
```

@kjk 
Hmmm why so fast ?  the stress test is looking more random than older test, perhaps it has changed to less actions ? so am switching to page range say 1-5
 
@laspinacristian
 this shows if you replace with filename a few pages (I use 1-5) the time of local opening remote from command line with some timers
running from cmd in exe dir 

```
time <nul & echo SumatraPDF.exe -exit-when-done -stress-test "remote file\filename.pdf" 1-5 > %temp%\temp.cmd &&echo "cmd /r time <nul" >>%temp%\temp.cmd &&call %temp%\temp.cmd
```

3.1.2
The current time is: 21:11:34.53
The current time is: 21:11:35.46
3.5.2
The current time is: 21:15:52.62
The current time is: 21:15:54.09

both under 2 seconds on a windows mapped remote drive (v:\share)

**Reply by laspinacristian** [reply]
I Will try these test too later.
What protocol Is your mapped drive using?

**Reply by GitHubRulesOK** [reply]
My Drive is a very very slow atom netbook (tiny screen) that has no battery and half its keys missing so trashed by the childers and should have been binned years ago (it thinks it was built in 2009 but I think its older as it was XP in a former life) but has 2gb memory and runs 24h/Win7 x32 for months at a time without reboot so a very good candidate for windows file sharing. As its on slowest network speeds, not used for high speed streaming, but good for basic MP4 Media files access, so Wildlife.wmv (30FPS) is smooth without stutter.
